src/Lykke.AlgoStore.Services/Validation/CSharpCodeBuildSession.cs

```diff
@@ -0,0 +1,296 @@
+                        .GetProperties(BindingFlags.Instance | BindingFlags.Public |
+                                       BindingFlags.GetProperty | BindingFlags.SetProperty);
+
+                    foreach (var functionParameter in functionParameters)
+                    {
+                        var parameter = ToAlgoMetaDataParameter(functionParameter);
+
+                        if (functionParameter.PropertyType.IsEnum)
+                            parameter.PredefinedValues = ToEnumValues(functionParameter);
+
+                        function.Parameters.Add(parameter);
+                    }
+
+                    metadata.Functions.Add(function);
+                }
+            }
+
+            return metadata;
+        }
+
+        private ValidationMessage DiagnosticToValidationMessage(Diagnostic diagnostic)
+        {
+            var validationMessage = new ValidationMessage();
+
+            var position = diagnostic.Location.GetLineSpan().StartLinePosition;
+
+            validationMessage.Line = (uint) position.Line;
+            validationMessage.Column = (uint) position.Character;
+
+            validationMessage.Id = diagnostic.Id;
+            validationMessage.Message = diagnostic.ToString();
+
+            switch (diagnostic.Severity)
+            {
+                case DiagnosticSeverity.Error:
+                    validationMessage.Severity = ValidationSeverity.Error;
+                    break;
+
+                case DiagnosticSeverity.Warning:
+                    validationMessage.Severity = ValidationSeverity.Warning;
+                    break;
+
+                default:
+                    validationMessage.Severity = ValidationSeverity.Info;
+                    break;
+            }
+
+            return validationMessage;
+        }
+
+        private bool ErrorExists(IEnumerable<ValidationMessage> messages)
+        {
+            return messages.Any(v => v.Severity == ValidationSeverity.Error);
+        }
+
+        private ValidationResult CreateAndSetValidationResult(
+            out ValidationResult validationResult,
+            bool isSuccessful,
+            IReadOnlyCollection<ValidationMessage> messages)
+        {
+            validationResult = new ValidationResult(isSuccessful, messages);
+
+            return validationResult;
+        }
+
+        private Assembly GenerateAssembly()
+        {
+            using (var ms = new MemoryStream())
+            {
+                //Emit results into a stream
+                var emitResult = _compilation.Emit(ms);
+
+                if (!emitResult.Success)
+                {
+                    // if not successful, throw an exception
+                    var failures = emitResult.Diagnostics
+                        .Where(x => x.IsWarningAsError || x.Severity == DiagnosticSeverity.Error);
+
+                    var message = string.Join(Environment.NewLine, failures.Select(x => $"{x.Id}: {x.GetMessage()}"));
```

Comment:
This probably won't include the location in the code where this error occured. Consider going through the `ValidationMessage` class to easily add it into the message. For example, `x => DiagnosticToValidationMessage(x).ToString()`